Port overview

attiks opened the pull request:
- [x] Code ported
- [x] Code linted
- [x] Tested

Refs: #RW-606

attiks committed:
Partial port

attiks committed:
Code ported and linted

attiks committed:
Fix package name

attiks committed:
Use organization from config

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+
+  this.overviewBoard = null;
+  this.boards = null;
+  this.projects = null;
```
This should be initialized to `{}` or `new Map()`. (The latter may ease iteration and lookup).

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+  this.trelloClient = trelloClient;
+
+  this.overviewBoard = null;
+  this.boards = null;
```
This should be initialized to `[]`.

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+   */
+  this.process = async () => {
+    try {
+      if (this.getBoards()) {
```
```suggestion
      if (await this.getBoards()) {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+   * Retrieve organization boards.
+   */
+  this.getBoards = async () => {
+    const boards = await this.trelloClient.get('/organizations/' + this.config.organization + '/boards', {
```
```suggestion
    const boards = await this.trelloClient.get('/organizations/' + this.config.trello.organization + '/boards', {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    }
+
+    try {
+      board.cards = await this.trelloClient.get('/boards/' + board.id, data);
```
```suggestion
      board.cards = await this.trelloClient.get('/boards/' + board.id + '/cards', data);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    };
+
+    if (checklists) {
+      date['checklists'] = 'all';
```
```suggestion
      data['checklists'] = 'all';
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    });
+
+    for (let board of boards) {
+      if (board.id == this.config.boardId) {
```
```suggestion
      if (board.id == this.config.trello.boardId) {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+
+    for (let board of boards) {
+      if (board.id == this.config.boardId) {
+        this.getBoardData(board, false);
```
```suggestion
        await this.getBoardData(board, true);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+        this.overviewBoard = board;
+      }
+      else {
+        this.getBoardData(board, false);
```
```suggestion
        await this.getBoardData(board, false);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    for (const project of this.projects) {
+      let card = project.card;
+      for (const checklist of card.checklist) {
+        if (boards[checklist.name]) {
```
```suggestion
        if (boards.has(checklist.name)) {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+      let card = project.card;
+      for (const checklist of card.checklist) {
+        if (boards[checklist.name]) {
+          let actions = project.actions[checklist.name];
```
```suggestion
          let actions = project.actions.get(checklist.name);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+        if (boards[checklist.name]) {
+          let actions = project.actions[checklist.name];
+          if (actions) {
+            this.updateChecklist(card.id, checklist, actions);
```
```suggestion
            await this.updateChecklist(card.id, checklist, actions);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+            project.actions.delete(checklist.name);
+          }
+          else {
+            this.deleteChecklist(checklist.id);
```
```suggestion
            await this.deleteChecklist(checklist.id);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+          }
+        }
+        else {
+          this.deleteChecklist(checklist.id);
```
```suggestion
          await this.deleteChecklist(checklist.id);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+      }
+
+      for (const [checklistName, actions] of project.actions) {
+        this.addChecklist(card.id, checklistName, actions);
```
```suggestion
        await this.addChecklist(card.id, checklistName, actions);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+        if (items.has(checkitem.name)) {
+          let action = items.get(checkitem.name);
+          if (action.name != checkitem.name) {
+            this.updateCheckItem(cardId, checklist.id, checkitem.id, action);
```
```suggestion
           await this.updateCheckItem(cardId, checklist.id, checkitem.id, action);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+        }
+      }
+      else {
+        this.deleteCheckItem(checklist.id, checkitem.id);
```
```suggestion
        await this.deleteCheckItem(checklist.id, checkitem.id);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    }
+
+    for (const action of items) {
+      this.addCheckItem(checklist.id, action);
```
```suggestion
      await this.addCheckItem(checklist.id, action);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+      this.logger.debug('Added checklist ' + name);
+
+      for (const action of actions) {
+        this.addCheckItem(checklistId, action);
```
```suggestion
        await this.addCheckItem(checklistId, action);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+   * Prepare the data to update the overview board.
+   */
+  this.prepare = async () => {
+    let projects = this.projects;
```
```suggestion
    let projects = this.getProjects();
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+        let actions = this.getActions(card, lists[card.idList]);
+        for (const [name, action] of actions) {
+          if (projects[name]) {
+            if (!projects.actions[board.name]) {
```
```suggestion
            if (!projects.actions.has(board.name)) {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+        for (const [name, action] of actions) {
+          if (projects[name]) {
+            if (!projects.actions[board.name]) {
+              projects.actions[board.name] = [];
```
```suggestion
              projects.actions.set(board.name, []);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+            if (!projects.actions[board.name]) {
+              projects.actions[board.name] = [];
+            }
+            projects.actions[board.name].push(action);
```
```suggestion
            projects.actions.get(board.name).push(action);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+        if (projectName != '') {
+          this.projects[projectName] = {
+            card: card,
+            actions: []
```
```suggestion
            actions: new Map(),
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+  /**
+   * Get Overview board projects.
+   */
+  this.getProjects = async () => {
```
```suggestion
  this.getProjects = () => {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+  /**
+   * Prepare the data to update the overview board.
+   */
+  this.prepare = async () => {
```
```suggestion
  this.prepare = () => {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+  /**
+   * Get the projects actions for a card.
+   */
+  this.getActions = async (card, list) => {
```
```suggestion
  this.getActions = (card, list) => {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    for (const label of card.labels) {
+      let projectName = this.getProjectName(label.name);
+      if (projectName != '') {
+        actions[projectName] = {
+          link: card.ShortUrl,
+          status: list,
+          due: card.Due,
+          complete: false,
+        };
```
```suggestion
        actions.set(projectName, {
          link: card.ShortUrl,
          status: list,
          due: card.Due,
          complete: false,
        });
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+  /**
+   * Compute an action name.
+   */
+  this.getActionName = async (action) => {
```
```suggestion
  this.getActionName = (action) => {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+   */
+  this.getActionName = async (action) => {
+    let doers = '*Not assigned*';
+    if (action.Doers.length > 0) {
```
```suggestion
    if (action.doers.length > 0) {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+  this.getActionName = async (action) => {
+    let doers = '*Not assigned*';
+    if (action.Doers.length > 0) {
+      doers = '**' + action.Doers.join(', ') + '**';
```
```suggestion
      doers = '**' + action.doers.join(', ') + '**';
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    }
+
+    let parts = [
+      action.Link,
```
```suggestion
      action.link,
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    let parts = [
+      action.Link,
+      doers,
+      action.Status
```
```suggestion
      action.status
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+      action.Status
+    ];
+
+    if (action.Due != '') {
```
```suggestion
    if (action.due != '') {
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+    ];
+
+    if (action.Due != '') {
+      parts.push(parts, '*' + action.Due + '*');
```
```suggestion
      parts.push(parts, '*' + action.due + '*');
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+   * Get doer name.
+   */
+  this.getDoerName = async (label) => {
+    if (label.indexOf(this.config.doerPrefix) === 0) {
+      return label.substring(this.config.doerPrefix.length);
```
```suggestion
    if (label.indexOf(this.config.trello.doerPrefix) === 0) {
      return label.substring(this.config.trello.doerPrefix.length);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+   * Get project name.
+   */
+  this.getProjectName = async (label) => {
+    if (label.indexOf(this.config.projectPrefix) === 0) {
+      return label.substring(this.config.projectPrefix.length);
```
```suggestion
    if (label.indexOf(this.config.trello.projectPrefix) === 0) {
      return label.substring(this.config.trello.projectPrefix.length);
```

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,360 @@
+  this.process = async () => {
+    try {
+      if (this.getBoards()) {
+        await this.prepare();
```
```suggestion
        this.prepare();
```

orakili reviewed (commented):
Sorry for the noise, just went through the code and marked what jumped at me. Probably missed some and some suggestions may need to be confirmed.

This app is indeed difficult to test. Maybe the setup could be reproduced by creating 2-3 private boards in your account, one being the "overview" board and then creating lists and cards with the `Doer: XXX` and `Project: YYY` labels.

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Update src/overview.js

Co-authored-by: orakili <github@orakili.net>

attiks committed:
Working code

attiks committed:
CS

attiks commented:
- Working code
- Each project can have a maximum of 1 card?
- ~~Removing the Id from the `for ... of` loops breaks the code~~

attiks committed:
Use values() for loops

attiks committed:
Use values() for loops

orakili commented:
> * Each project can have a maximum of 1 card?

Yes, normally (i.e if people follow the instructions correctly), there's supposed to be only 1 card with a given `Project: YYY` label on the overview board.

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,385 @@
+    let items = new Map();
+    for (const action of actions.values()) {
+      this.logger.debug('Processing action ' + action.name);
+      items.set(action.name, action);
```
This should use the `action.link` as key as it's the only non changing part of the check list items ("link - does - status - due"). See next comment below.

attiks replied:
Changed

orakili commented on src/overview.js:
```diff
@@ -0,0 +1,385 @@
+    }
+
+    for (const checkitem of checklist.checkItems) {
+      if (items.has(checkitem.name)) {
```
This should extract the link from the check list item name ("link - does - status - due") to do the lookup in the `items` map.

`action.name` may indeed differ from `checkitem.name` (previous value), for example if the status, doer or due changed. Only the beginning with the link is the same and serves as identifier.

Ref: https://github.com/UN-OCHA/legacy--rw-trello-overview/blob/fe4a507e53bf6d3fc28c58c5d3c064e5020bfa65/main.go#L231-L238

attiks replied:
Changed, but :100: % not sure

orakili reviewed (changes requested):
Looks good but the existing check list items are not updated properly when something changed like doer, status (card list) or due date.

attiks committed:
use action link

orakili committed:
fix: date format for due date

Refs: RW-606

orakili committed:
fix: edge case where the list couldn't be found

Refs: RW-606

orakili committed:
fix: request method to update check list item + spacing

Refs: RW-606

orakili committed:
fix: check list update logic to look up existing items

Refs: RW-606

orakili committed:
refactor: add list of boards to exclude and retrieve overview board sepately

This allows to use an overview board outside of the organization which is helpful for testing.

Refs: RW-606
